rotkehlchen/exchanges/okx.py

```diff
@@ -0,0 +1,410 @@
+            params.update({
+                'limit': options.get('limit', ''),
+                'after': options.get('after', ''),
+            })
+        elif endpoint == 'withdrawals':
+            path = '/api/v5/asset/withdrawal-history'
+            # supports pagination
+            params.update({
+                'limit': options.get('limit', ''),
+                'after': options.get('after', ''),
+            })
+        else:
+            raise AssertionError(f'Unexpected {self.name} endpoint type: {endpoint}')
+
+        if len(params) != 0:
+            path += f'?{urlencode(params)}'
+
+        timestamp = datetime.datetime.utcnow().isoformat(timespec='milliseconds') + 'Z'
+        signature = self._generate_signature(timestamp, method, path, '')
+
+        self.session.headers.update({
+            'OK-ACCESS-TIMESTAMP': timestamp,
+            'OK-ACCESS-SIGN': signature,
+        })
+
+        url = urljoin(self.base_uri, path)
+
+        try:
+            response = self.session.request(method=method, url=url)
+        except requests.exceptions.RequestException as e:
+            raise RemoteError(f'{self.name} API request failed due to {str(e)}') from e
+        try:
+            json_response = json.loads(response.text)
+        except JSONDecodeError as e:
+            raise RemoteError(
+                f'{self.name} returned invalid JSON response: {response.text}',
+            ) from e
+
+        return json_response
+
+    def _paginated_api_query(
+        self,
+        endpoint: Literal['currencies', 'balance', 'trades', 'deposits', 'withdrawals'],
+        pagination_key: str,
+        options: Optional[dict] = None,
+    ) -> list:
+        options = options.copy() if options else {}
+        all_items = []
+
+        while True:
+            response = self._api_query(endpoint, options)
+
+            if 'data' not in response:
+                raise RemoteError(f'{self.name} json response does not contain `data`: {response}')
+
+            all_items.extend(response['data'])
+
+            if len(response['data']) < self.MAX_RESULTS:
+                break
+
+            earliest_item = response['data'][-1]
+            # after: return items with pagination_key value "earlier" than this value
+            options.update({'after': earliest_item[pagination_key]})
+
+        return all_items
+
+    def _query_currencies(self) -> dict:
+        return self._api_query('currencies')
+
+    def validate_api_key(self) -> tuple[bool, str]:
+        response = self._api_query('balance')
+        if response.get("code", "") != "0":
+            return False, 'Error validating API credentials'
+        return True, ''
+
+    def query_balances(self, **kwargs: Any) -> ExchangeQueryBalances:
+        """
+        https://www.okx.com/docs-v5/en/#rest-api-account-get-balance
+        """
+        response = self._api_query('balance')
+
+        assets_balance: DefaultDict[AssetWithOracles, Balance] = defaultdict(Balance)
```

Comment:
why do you need a defaultdict instead of using a dict?

Reply:
Copied the pattern used elsewhere, but I believe using a dict here throws a `KeyError` on this line below:

```py
assets_balance[asset] += Balance(
```
since this is the only way the dictionary is populated and it expects the key (or a default value) to exist.